src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs

```diff
@@ -24,18 +24,7 @@ protected AbstractQualifyMemberAccessDiagnosticAnalyzer()
         {
         }
 
-        public override bool OpenFileOnly(Workspace workspace)
-        {
-            var qualifyFieldAccessOption = workspace.Options.GetOption(CodeStyleOptions.QualifyFieldAccess, GetLanguageName()).Notification;
```

Comment:
interesting there were so many open file only analyzers in IDE.. is it due to some hack IDE did like returning nothing for supported diagnsotics and then report some random diagnostics when they want causing engine's optimization of not running hidden diagnostics for closed files?

is that now changed?

Reply:
Sorry, can you explain why that's a hack?  How else would you get fix-all to work across documents otherwise?

Can you explain what the right way is to do things so the IDE can remove 'hacks'?    Thanks!

Reply:
> Can you explain what the right way is to do things so the IDE can remove 'hacks'? Thanks!

Reply:
Basically, what is the "right" way for the IDE to use the analyzer subsystem for things like fixes/style/etc?

Reply:
> it might have been better if IDE didn't jam these into compiler DIagnosticAnalyzer but built its own DiagnosticAnalyzer system for IDE only that can use whatever system they want for whatever experience they want rather than jam these conflicting desire into one that defined in compiler layer.

I'm confused.  The original design of the Diagnostic Analysis subsystem was to make something that would work for IDE purposes**.  IDE *already* had it's own system prior to that (after all, we shipped analyzers/fixers long before we build the Diagnostic-Analyzer stuff).  If DiagnosticAnalyzer isn't the appropriate system for the IDE then why was so much effort put into building it *and* moving all the existing IDE code onto it?  

--

** I remember this distinctly in the original meetings on the topic.  When we were deciding core deliverables, one of them was that this has to suffice for IDE needs since a primary goal was to get us out of hte "let's analyze files twice" state we were in.  So we went with: we'll analyze stuff once as hte compiler is doing it, and it will call out into plugins so that we can do things like IDE analysis at the same time.  This was a core goal, and i'm def confused by the messaging that the IDE was actually supposed to *not* use the actual DiagnosticAnalyzer...

Reply:
we can bring in @mavasani and @Pilchie since I wasn't in the design meeting that decide DiagnosticAnalyzer compiler feature. 

anyway, DiagnosticAnalyzer was built as compiler feature and from the point of view of compiler. that's why it doesn't have concept of file by file analysis concept which is very important for IDE.

and I dont believe it was built for all the possible things IDE could do in mind rather it was built with what all the thing compiler can do in mind. and IDE simply wants more than that. 

that's why DiagnosticAnalyzer can only access compilation data and nothing else when IDE can do much more with other data (ex, Nuget, web config, options and more)

as I remember, at the time, once it moved to compiler layer, it was simply errors that compiler produce for IDE that we shouldn't care whether it is csc generated or third party analyzers. and that's it.

I don't know why IDE decide to hop on that system. if they wanted some part of it like squiggle, LB, error list, you guys could use incremental analyzer, diagnostic source, and expand LB to support IDE one as well. since IDE DiagnosticAnalyzer can't run on VBCSCompiler anyway.

Reply:
or you guys could change compiler DiagnosticAnalyzer itself to support all things you guys wanted and then IDE side naturally get all those new thing working. rather than workaround these thing in IDE and now we have 2 different systems that share some and not share some.

Reply:
>  This was a core goal, and i'm def confused by the messaging that the IDE was actually supposed to not use the actual DiagnosticAnalyzer

if you guys were using DiagnosticAnalyzer as it is defined/supported by compiler then it was right decision. but you guys don't. and that is the problem.

Reply:
> I don't know why IDE decide to hop on that system. 

Literally a core design goal of DiagnosticAnalyzers was to be able to support the IDE scenarios around lightbulbs.  That was a design goal from the very earliest meetings.  It's one of the reasons that so many IDE features immediately moved onto DiagnosticAnalyzers.  Which i *though* was actually done by the analyzer team.  i.e. i don't think IDE people moved things like 'simplify name', 'remove-cast', 'generate method' and so on and so forth to DiagnosticAnalyzers.  That was done by the analyzer group to help prove out the DiagnosticAnalyzer infrastructure and to show that it could suffice for IDE needs.

If this wasn't the purpose of DiagnosticAnalyzer, then there was some *dramatic* miscommunication.  Because literally tons of man-hours were sunk to create this system and then have pretty much eveyr IDE diagnostic feature now sit on it.  If that wasn't the purpose of this work, then i have no idea why that wasn't noticed or addressed until now...

> or when IDE needed more, if IDE improved DiagnsoticAnalyzer in compiler then, it would have been good again since we could just call compiler.GetDiagnostics without any IDE only hacks and special behavior. but IDE doesn't. and that is the problem.

You're saying the IDE hacked things.  But it's unclear to me what those hacks actually are.  As far as i can tell, we're using the API that is actually exposed by the analysis subsystem.  And, from previous posts you said things like "IBuiltInAnalyzer interface is basically a hack to make 2 system merged together."  But IDE didn't create that API.  That API is what came from people working on the DiagnosticAnalyzer and it was added to the analyzer that the IDE had built.

--

Basically, if there was an issue, i wish it had been brought up 5+ years ago as we started using these facilities.  I'm happy to move to a 'non-hack' solution.  But it's unclear to me what that is. 

However, i don't think ti's a solution for the IDE to just make a new analysis subsystem for itself.  That would be enormous duplication.  And would go against the intended goal of the real diagnostic analysis work (which was to create something suitable for all these needs).  So any real solution would be updating the real DiagnosticAnalyzer to do things non-hackily.  However, that works seems like it would be sensible to do by the people who know the DiagnosticAnalyzer subsystem the best.

Reply:
DiagnosticAnalyzer API was always designed to support analyzers in both compiler and IDE hosts. However, due to various functional and/or performance requirements of executing these analyzers in the IDE and configuring them via workspace options, over time lot of workarounds have been added in the analyzers and the IDE diagnostic engine that strongly couple them to the IDE host. @sharwell's planned work on moving these analyzers to NuGet would thread apart all such tight couplings, will need removing all such workarounds and internals knowledge, and possibly identify missing public APIs or engine functionality that are required to continue executing these analyzers correctly and efficiently in IDE.

Reply:
> @sharwell's planned work on moving these analyzers to NuGet would thread apart all such tight couplings, will need removing all such workarounds and internals knowledge, and possibly identify missing public APIs or engine functionality that are required to continue executing these analyzers correctly and efficiently in IDE.

💯 The NuGet distribution of what is currently the IDE analyzers will not have `InternalsVisibleTo` access to the compiler or workspace, so whatever we need implemented to make the experience work well for those will necessarily be available for all analyzer authors.

Reply:
➡️ The changes to `OpenFileOnly` methods were reverted (no other changes so I amended the previous commit)